subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+If you are writing convention plugins, this is usually not an issue. But if you distribute your plugins, you should pay attention to their compatibility.
+
+[[sec:gradle_api]]
+=== The Gradle API
+
+All plugins use the https://docs.gradle.org/current/javadoc/index.html[Gradle Core API] to configure a Gradle build. The Gradle API must be added as a build compile-time dependency using the helper function `gradleApi()`
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly(gradleApi())
+}
+----
+
+TIP: `gradleApi()` must be a `compileOnly` dependency, because the correct Gradle API will be provided by Gradle at runtime.
+
+==== Version of the Gradle API Version
+
+The Gradle API version provided by `gradleApi()` is the same as the version of Gradle used to build the plugin.
+
+For example, if the plugin is built using Gradle version 7.6.1, then `gradleApi()` also be version 7.6.1. If the plugin is later built with Gradle {gradleVersion}, then the `gradleApi()` will change to be version {gradleVersion}.
+
+This implicit link between affects the Gradle versions your plugin supports. It's easy to build a plugin using one version of Gradle that that has functions or classes that are not present in other versions of Gradle.
+
+This might result in runtime exceptions, such as `NoSuchMethodError`, if a plugin uses a functionality of the Gradle API that is missing in the Gradle being used to run the plugin. This might happen the plugin is run by an older Gradle version that is missing a function or class, _or_ if the plugin uses a function or class that has been removed in newer versions (which can happen when using deprecated or internal Gradle features).
+
+Supporting different versions of Gradle can be tested using https://docs.gradle.org/current/userguide/test_kit.html#sub:gradle-runner-gradle-version[TestKit].
+
+[[sec:redistributed_gradle_api]]
+=== Redistributed Gradle API
+
+To avoid the implicit link on a Gradle version that comes with using `gradleApi()`, an alternative is to use the https://docs.nokee.dev/manual/gradle-plugin-development.html[Nokee redistributed Gradle API artifacts]. These are like `gradleApi()` except they allow specifying a specific version of the Gradle API.
+
+[subs="attributes"]
+[source,kotlin]
+----
+dependencies {
+    compileOnly("dev.gradleplugins:gradle-api:{gradleVersion}")
+}
+----
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+This makes the version of the Gradle API being used to build the plugin more explicit, removing the ambiguity avoids the implicit link between `gradleApi()` and the build version of Gradle.
+
+On the other hand, you cannot use newer APIs if your plugin is running on a recent version of Gradle.
```

Comment:
🔍 With the redistributed Gradle API, you can create multiple source sets, each targeting a different Gradle API. Solving the issue exposed here. However, all the wiring needs to be done by hand. We hope to offer a convenient solution soon in the [Toolbox project](https://github.com/gradle-plugins/toolbox).